test/hotspot/jtreg/compiler/c2/irTests/TestMinMaxIdentities.java

```diff
@@ -0,0 +1,192 @@
+        assertResult(RANDOM.nextInt(), RANDOM.nextInt(), RANDOM.nextLong(), RANDOM.nextLong(), RANDOM.nextFloat(), RANDOM.nextFloat(), RANDOM.nextDouble(), RANDOM.nextDouble());
+
+        assertResult(Integer.MAX_VALUE, Integer.MIN_VALUE, Long.MAX_VALUE, Long.MIN_VALUE, Float.POSITIVE_INFINITY, Float.NaN, Double.POSITIVE_INFINITY, Double.NaN);
+        assertResult(Integer.MIN_VALUE, Integer.MAX_VALUE, Long.MIN_VALUE, Long.MAX_VALUE, Float.NaN, Float.POSITIVE_INFINITY, Double.NaN, Double.POSITIVE_INFINITY);
+    }
+
+    @DontCompile
+    public void assertResult(int iA, int iB, long lA, long lB, float fA, float fB, double dA, double dB) {
+        Asserts.assertEQ(Math.min(iA, Math.min(iA, iB)), intMinMin(iA, iB));
+        Asserts.assertEQ(Math.min(iA, Math.max(iA, iB)), intMinMax(iA, iB));
+        Asserts.assertEQ(Math.max(iA, Math.min(iA, iB)), intMaxMin(iA, iB));
+        Asserts.assertEQ(Math.max(iA, Math.max(iA, iB)), intMaxMax(iA, iB));
+
+        Asserts.assertEQ(Math.min(lA, Math.min(lA, lB)), longMinMin(lA, lB));
+        Asserts.assertEQ(Math.min(lA, Math.max(lA, lB)), longMinMax(lA, lB));
+        Asserts.assertEQ(Math.max(lA, Math.min(lA, lB)), longMaxMin(lA, lB));
+        Asserts.assertEQ(Math.max(lA, Math.max(lA, lB)), longMaxMax(lA, lB));
+
+        Asserts.assertEQ(Math.min(fA, Math.min(fA, fB)), floatMinMin(fA, fB));
+        Asserts.assertEQ(Math.max(fA, Math.max(fA, fB)), floatMaxMax(fA, fB));
+
+        Asserts.assertEQ(Math.min(dA, Math.min(dA, dB)), doubleMinMin(dA, dB));
+        Asserts.assertEQ(Math.max(dA, Math.max(dA, dB)), doubleMaxMax(dA, dB));
+
+        // Due to NaN, these identities cannot be simplified.
+
+        Asserts.assertEQ(Math.min(fA, Math.max(fA, fB)), floatMinMax(fA, fB));
+        Asserts.assertEQ(Math.max(fA, Math.min(fA, fB)), floatMaxMin(fA, fB));
+        Asserts.assertEQ(Math.min(dA, Math.max(dA, dB)), doubleMinMax(dA, dB));
+        Asserts.assertEQ(Math.max(dA, Math.min(dA, dB)), doubleMaxMin(dA, dB));
+    }
+
+    // Integers
+
+    @Test
+    @IR(counts = { IRNode.MIN_I, "1" })
+    public int intMinMin(int a, int b) {
+        return Math.min(a, Math.min(a, b));
+    }
+
+    @Test
+    @IR(failOn = { IRNode.MIN_I, IRNode.MAX_I })
+    public int intMinMax(int a, int b) {
+        return Math.min(a, Math.max(a, b));
+    }
+
+    @Test
+    @IR(failOn = { IRNode.MIN_I, IRNode.MAX_I })
+    public int intMaxMin(int a, int b) {
+        return Math.max(a, Math.min(a, b));
+    }
+
+    @Test
+    @IR(counts = { IRNode.MAX_I, "1" })
+    public int intMaxMax(int a, int b) {
+        return Math.max(a, Math.max(a, b));
+    }
+
+    // Longs
+
+    @Test
+    @IR(applyIfPlatform = { "riscv64", "false" }, phase = { CompilePhase.BEFORE_MACRO_EXPANSION }, counts = { IRNode.MIN_L, "1" })
+    public long longMinMin(long a, long b) {
+        return Math.min(a, Math.min(a, b));
+    }
+
+    @Test
+    @IR(applyIfPlatform = { "riscv64", "false" }, failOn = { IRNode.MIN_L, IRNode.MAX_L })
```

Comment:
Since `MinL/MaxL` are expanded in macro expansion, this rule will also succeed even if the optimization is not applied. I suggest to also add `phase = CompilePhase.BEFORE_MACRO_EXPANSION`. Same below.